test/e2e/federate.go

```diff
@@ -0,0 +1,148 @@
+				if err != nil {
+					tl.Fatalf("Error retrieving federatedtypeconfig %q: %v", typeConfigName, err)
+				}
+
+				if framework.TestContext.LimitedScope && !typeConfig.GetNamespaced() {
+					framework.Skipf("Federation of cluster-scoped type %s is not supported by a namespaced control plane.", typeConfigName)
+				}
+
+				kind := typeConfig.GetTarget().Kind
+				targetAPIResource := typeConfig.GetTarget()
+				targetObject, err := common.NewTestTargetObject(typeConfig, f.TestNamespaceName(), fixture)
+				if err != nil {
+					tl.Fatalf("Error creating test objects: %v", err)
+				}
+
+				createdTargetObject, err := common.CreateResource(kubeconfig, targetAPIResource, targetObject)
+				if err != nil {
+					tl.Fatalf("Error creating resource: %v", err)
+				}
+
+				qualifiedTypeName := util.QualifiedName{
+					Name:      typeConfig.GetObjectMeta().Name,
+					Namespace: typeConfig.GetObjectMeta().Namespace,
+				}
+				qualifiedTargetName := util.NewQualifiedName(createdTargetObject)
+
+				tl.Logf("Federating %s %q", kind, qualifiedTargetName)
+				federatedObject, err := federate.FederateResource(kubeconfig, qualifiedTypeName, qualifiedTargetName, false)
+				if err != nil {
+					tl.Fatalf("Error federating %s %q: %v", kind, qualifiedTargetName, err)
+				}
+
+				validateTemplateEquality(tl, federatedObject, createdTargetObject, typeConfig.GetFederatedType().Kind)
+
+				fedAPIResource := typeConfig.GetFederatedType()
+				deleteResources(tl, kubeconfig, qualifiedTargetName, fedAPIResource)
+			})
+		})
+	}
+})
+
+func validateTemplateEquality(tl common.TestLogger, fedObj, targetObj *unstructured.Unstructured, fedKind string) {
+	qualifiedName := util.NewQualifiedName(fedObj)
+	templateMap, ok, err := unstructured.NestedFieldCopy(fedObj.Object, util.SpecField, util.TemplateField)
+	if err != nil || !ok {
+		tl.Fatalf("Error retrieving template from %s %q", fedKind, qualifiedName)
+	}
+
+	expectedObj := &unstructured.Unstructured{}
+	expectedObj.Object = templateMap.(map[string]interface{})
+	federate.RemoveUnwantedFields(expectedObj)
+	federate.RemoveUnwantedFields(targetObj)
+
+	if !reflect.DeepEqual(expectedObj, targetObj) {
+		tl.Fatal("Federated object template and target object don't match for %s %q", fedKind, qualifiedName)
+	}
+}
+
+func deleteResources(tl common.TestLogger, kubeconfig *restclient.Config, qualifiedName util.QualifiedName, apiResource metav1.APIResource) {
+	fedKind := apiResource.Kind
+	client, err := util.NewResourceClient(kubeconfig, &apiResource)
+	if err != nil {
+		tl.Fatalf("Error getting resource client for %s", fedKind)
+	}
+
+	tl.Logf("Deleting %s %q", fedKind, qualifiedName)
+	// orphanDependents ensures the resources propagated to clusters are also deleted
+	orphanDependents := false
+	err = client.Resources(qualifiedName.Namespace).Delete(qualifiedName.Name, &metav1.DeleteOptions{OrphanDependents: &orphanDependents})
```

Comment:
There is a wait loop to check that the resource is gone after delete is called. Did you mean delete itself called in a wait loop in case it returns failure, until the failure == `notFound`?

Reply:
I think performing the deletion in a wait loop as well would minimize the potential for transient failure.  Just a thought, though, there is a lot of existing code that isn't this rigorous.